File: RayTracerAvalonia/RayTracing/Ray.cs
using RayTracerAvalonia.RayTracing.Extensions;
using System;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;
using RayTracerAvalonia.RayTracing.Shapes;

namespace RayTracerAvalonia.RayTracing;
public record struct Ray
{
    public Vector3 Start { get; }
    public Vector3 Direction { get; }

    public Ray(Vector3 start, Vector3 direction)
    {
        Start = start;
        Direction = direction.Normalize();
    }
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public Color Trace(Scene scene)
    {
        float? minDistance = null;
        IShape? closestShape = null;
        Vector3 point = default;

        foreach (var shape in scene.Shapes)
        {
            var distance = IShape.ClosestDistanceAlongRay(shape, this);
            if (distance is null) continue;
            if (minDistance is not null && !(distance < minDistance)) continue;
            minDistance = distance;
            closestShape = shape;
            point = Start + (Direction * minDistance.Value);
        }
        return closestShape?.GetColorAt(point, scene) ?? scene.BackgroundMaterial.GetColorAt(default);
    }
}

File: RayTracerAvalonia.Desktop/cpp_raytracing_engine/ray.cpp
#include "include/ray.hpp"
#include "include/color.hpp"
#include "include/scene.hpp"
#include <Eigen/Core>
#include <memory>

using Eigen::Vector3f;

// Implementation of Ray methods

Color Ray::trace(const Scene &scene) const {
  // TODO: Implement intersection logic and shading
  // For now, return background color as a placeholder
  auto minDistance = std::numeric_limits<float>::infinity();
  Shape *closestShape = nullptr;
  Vector3f point;

  for (const auto &shape : scene.shapes) {

    auto distance = shape->ClosestDistanceAlongRay(*this);
    if (!distance.has_value())
      continue;
    if (*distance < minDistance) {

      minDistance = *distance;
      closestShape = shape.get();
      point = this->start + (this->direction * minDistance);
    }
  }

  return closestShape != nullptr ? closestShape->getColorAt(point, scene)
                                 : scene.background;
}
